elastic/kibana
# [AO] Compile the KQL expression on the server

Closes #159017

## Summary

This PR is related to this [comment](https://github.com/elastic/kibana/pull/158665#discussion_r1212385677). Previously, we had both `fitlerQueryText` (unparsed KQL expression) and `filterQuery` (stringified JSON representation of the KQL expression). This makes the JSON body for creating a rule via the API confusing, as @simianhacker pointed out. In this PR, I've removed `filterQuery` and renamed `fitlerQueryText` to `filterQuery` to avoid confusion and parsed the `filterQuery` when needed.

After this change, you should only see `filterQuery` in the AAD and the rule definition with the user's value.

|Rule definition|Alert document|
|---|---|
|![image](https://github.com/elastic/kibana/assets/12370520/05fb04c1-4f5a-4587-b7ac-a90a088b5f26)|![image](https://github.com/elastic/kibana/assets/12370520/2471d144-f284-4721-94ec-cf1b95029786)|

## 🧪 How to test
- Create a new threshold rule and add a filter for it
- Make sure that all the rule functionalities that use filter work as expected
    - Preview
    - Rule execution
    - Saved data in rule definition and alert document
- Add an invalid filter query by editing the rule and make sure the validation works as expected (filter should be invalid state and save button should not work)
